nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+        assert!(twap.observations[1].price2_cumulative == U128(13), "updated price 2 Mismatch");
+    }
+
+    #[test]
+    fn overwrite_works() {
+        init_blockchain();
+
+        let mut twap: Twap = Twap::new();
+        let mut last_updated_index = twap.initialize(env::block_timestamp(), 1, 1);
+
+        let max_length = 10;
+        // fill all places
+        for i in 1..10 {
+            let timestamp = env::block_timestamp() + i;
+            last_updated_index = twap.write(
+                timestamp,
+                1, 1,
+                max_length
+            );
+        }
+
+        assert!(twap.observations.len() == 10, "Mismatch");
+
+        // next observation should be written on 0th Index
+        let mut last_timestamp = env::block_timestamp() + 10;
+        last_updated_index = twap.write(
+            last_timestamp,
+            1, 1,
+            max_length
+        );
+
+        assert!(twap.observations.len() == 10, "Mismatch");
+        assert!(twap.observations[0].block_timestamp == U64(last_timestamp), "Mismatch");
+        assert!(twap.observations[0].num_of_observations == U128(11));
+
+        // next observation should be written on 1st Index
+        last_timestamp = env::block_timestamp() + 11;
+        last_updated_index = twap.write(
+            last_timestamp,
+            1, 1,
+            max_length
+        );
+
+        env_log!("as {}", twap.observations.len());
+        assert!(twap.observations.len() == 10, "Mismatch");
+        assert!(last_updated_index == 1, "current index mismatch");
+
+        assert!(twap.observations[1].block_timestamp == U64(last_timestamp), "Mismatch");
+        assert!(twap.observations[1].num_of_observations == U128(12));
+    }
+
+    #[test]
+    fn simple_binary_search_works() {
+        init_blockchain();
+
+        let twap: Twap = get_twap();
+        let max_length = 10;
+
+        // current observation timestamp array [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+        let mut returned_index = twap.binary_search(
+            max_length,
+            5,
+        );
+
+        assert!(returned_index == 4, "Wrong Index");
+
+        returned_index = twap.binary_search(
+            max_length,
+            0,
+        );
+
+        assert!(returned_index == 0, "Wrong Index");
+
+        returned_index = twap.binary_search(
+            max_length,
+            10,
+        );
+
+        assert!(returned_index == 9, "Wrong Index");
+    }
+
+    #[test]
+    #[should_panic(expected = "Observation after this timestamp doesn't exist")]
+    fn binary_edge_case_works() {
+        init_blockchain();
+
+        let twap: Twap = get_twap();
+        let max_length = 10;
+
+        // current observation timestamp array [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+        twap.binary_search(
+            max_length,
+            122,
+        );
+    }
+
+    #[test]
+    fn pivoted_binary_search_works() {
+        init_blockchain();
```

Comment:
do we need to initialize blockchain here? We only need TWAP  structure.

Reply:
if i don't do this i get error, maybe this is because we are using env::block_timestamp for current block

Reply:
then let's put `env::block_timestamp` as a parameter. With this, the TWAP will be self contained and will not depend on the blockchain.